com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardFieldPropertyView.cs

```diff
@@ -1,18 +1,53 @@
 using System;
+using System.Linq;
 using UnityEditor.Experimental.UIElements;
 using UnityEditor.Graphing;
+using UnityEditor.Graphing.Util;
 using UnityEngine;
 using UnityEngine.Experimental.UIElements;
+using UnityEngine.Experimental.UIElements.StyleSheets;
+using Toggle = UnityEngine.Experimental.UIElements.Toggle;
 
 namespace UnityEditor.ShaderGraph.Drawing
 {
     class BlackboardFieldPropertyView : VisualElement
     {
         readonly AbstractMaterialGraph m_Graph;
 
+        IShaderProperty m_Property;
+        TextField m_ReferenceNameField;
+
+        static Type s_ContextualMenuManipulator = AppDomain.CurrentDomain.GetAssemblies().SelectMany(x => x.GetTypesOrNothing()).FirstOrDefault(t => t.FullName == "UnityEngine.Experimental.UIElements.ContextualMenuManipulator");
+
+        IManipulator m_ResetReferenceMenu;
+
         public BlackboardFieldPropertyView(AbstractMaterialGraph graph, IShaderProperty property)
         {
             m_Graph = graph;
+            m_Property = property;
+
+            m_ReferenceNameField = new TextField(512, false, false, ' ');
+            AddRow("Reference", m_ReferenceNameField);
+            m_ReferenceNameField.value = property.referenceName;
+            m_ReferenceNameField.isDelayed = true;
+            m_ReferenceNameField.OnValueChanged(newName =>
+            {
+                string newReferenceName = m_Graph.SanitizePropertyReferenceName(newName.newValue, property.guid);
+                property.overrideReferenceName = newReferenceName;
+                m_ReferenceNameField.value = property.referenceName;
+
+                if (string.IsNullOrEmpty(property.overrideReferenceName))
+                    m_ReferenceNameField.style.fontStyle = StyleValue<FontStyle>.Create(FontStyle.Normal);
```

Comment:
Not a fan of styles being inlined here, should use a USS class instead (e.g. `overriden`)